metadata

leoscholl opened the pull request:
see what you think about the way I've implemented the metadata structure. two files to look at:
- MetadataExtract.m opens the dataset if not already open, copies fields of interest to the result structure, the rest of the work is left for the GUI after all datasets in a batch have been exported
- MetaTable.m class that implements loading, appending, query, and saving the table

two things I'm mainly concerned with:
- will it be easy to add support for optical imaging data?
- are matlab string arrays (new since 2016a) ok to use? compatible with julia?

Commit:
don't load handles

some vis stim data contains handles and matlab tries to open them

Commit:
Merge branch 'master' into leo

Commit:
visstim condtestcond table

Commit:
add metadata storage

babaq commented on +NeuroAnalysis/+IO/Export.m:
```diff
@@ -10,39 +10,43 @@
-    copyfile(dataset,exportpath);
+    copyfile(dataset,filepath);
+    dataset = struct;
+    dataset.filepath = filepath;
```
i don't quite understand all the code change above this line. the result of Export function is mostly exception information. result.status indicate whether successfully prepare the dataset. the result.source is the full path of the file to be prepared. The Dataset, however is the prepared dataset which contain all the information needed to fully describe the dataset. dataset.filepath is the file path of the exported dataset mat file.

babaq commented on +NeuroAnalysis/+IO/Export.m:
```diff
@@ -10,39 +10,43 @@
 end
-result.status = true;
-result.source = filename;
 disp('Exporting Dataset:    Done.');
```
i don't quite understand all the code change above this line. the result of Export function is mostly exception information. result.status indicate whether successfully prepare the dataset. the result.source is the full path of the file to be prepared. The Dataset, however is the prepared dataset which contain all the information needed to fully describe the dataset. dataset.filepath is the file path of the exported dataset mat file. so after the data exported, the result should be ready containing exception information about the exporting process.

babaq commented on +NeuroAnalysis/+IO/Export.m:
```diff
@@ -10,39 +10,43 @@
 end
-
+%% Update metadata file
+result = NeuroAnalysis.Base.EvalFun('NeuroAnalysis.IO.MetadataExtract', {dataset,exportpath,sourceformat,isparallel,callbackresult});
```
i think it's better to separate meta data preparation into individual `NeuroAnalysis.sourceformat.PrepareMetaData` functions. because here is where each dataset have already exported and analyzed, in serial order or parallel. if we just extract aproparate metadata for each dataset and put it in result. we could merge them into the global metadata once after Export whether it is a batch or not. so i don't think we need to worry about `isparallel`, `sourceformat` etc in argument list. just `NeuroAnalysis.Base.EvalFun(['NeuroAnalysis.',sourceformat,'.PrepareMetaData'],[{dataset,callbackresult},varargin])`

leoscholl replied:
Ok. We still need a function for updating the metadata file in batch, without exporting. It can be separate, like `NeuroAnalysis.IO.ExportMetadata` or similar, that just opens individual datasets (in parallel or serial) and sends it to the appropriate `PrepareMetadata` function?

babaq commented on +NeuroAnalysis/+IO/MetadataExtract.m:
```diff
@@ -0,0 +1,57 @@
+function [ result ] = MetadataExtract( dataset,exportpath,sourceformat,isparallel,varargin )
```
i still think we should separate `PrepareMetaData` into different sourceformat, and here call those `PreareMetaData`. because we can't guarantee each dataset will generate metadata that have `subject`, `recordingsession` fields etc. for example, optical imaging block file doesn't contain any of there fields.

babaq commented:
i tried to read matlab string array, and it failed, the matlab string array is a matlab class, and all matlab class is saved in propertory HDF5 class type subsystem which julis don't know the format, hence not be able to read it.

babaq commented on +NeuroAnalysis/+IO/MetaTable.m:
```diff
@@ -0,0 +1,117 @@
+classdef MetaTable < handle
```
honestly, i am not a fan of matlab object-oriented things. i thought we only need an table, and some functions working on the table. but i see you added a map alongside the tests. can you elaborate why this map is needed?

Commit:
metadata in scalar structure, prepare separate from update

babaq committed:
update evalfun

babaq committed:
update prepare metadata

babaq committed:
function names are better start with verbs

babaq committed:
fuction name change

babaq committed:
function name change

since here is only collecting meta that are already exist in exported dataset, and the metadata collection is not export to any file, function name should be changed accordingly. the final compare and merge will be done in GUI `CollectMetadata` function.

leoscholl commented on +NeuroAnalysis/+IO/MetaTable.m:
```diff
@@ -0,0 +1,128 @@
+            % Merge according to sourceformat if necessary
+            toMerge = false(length(Tests),1);
+            for t = 1:length(Tests)
+                toMerge(t) = isequal(Tests(t).(test.key), test.(test.key));
```
I'm still not sure about this; i've required that the metadata prepare function identify one field as a 'key' that uniquely identifies each test. For ripple data, this would be the filename, but for optical imaging / vlab data you would need to think of something else to use. Also I'm not sure if we need to be checking if the sourceformat matches as well, e.g. could there be multiple tests with the same filename but different source formats?

babaq replied:
i think the logical way is to first search for relevant entries by filtering on properties such as `sourceformat`, `test`, `subject`, etc. if result is empty means this is a new entry to be added, if result is one entry then merge, if greater than one, warn, record ambiguous entries and skip, we could deal with them later.

Commit:
add metadata synchronize

Commit:
update gui, standardize param names

not validating param names or types yet

babaq committed:
minor update

change error and assert function to warning & return to provent interuption of processing, clear vlab and export funtion.

babaq committed:
preare metadata

reorganize ripple preparemetadata, defer experimental meta to vlab and visstim preparemetadata

babaq committed:
clean collectmetadata

babaq committed:
change test folder name

Commit:
visstim prepare changes

fix stimtimes adjustment for short TTL pluses, add visstim preparemetadata

Commit:
add standardize params

also fixed result.source being set if there is an error

Commit:
clean up metadata

change name addEntry -> addRow
gui has option to change metadata path
metadata merge no longer depends on date; newer file always appends older file
query now returns a new MetaTable object
require callback result to be a struct

babaq committed:
move stadardize ex into each prepare function, merge vlab getparam to base and some cleaning

babaq committed:
Revert "move stadardize ex into each prepare function, merge vlab getparam to base and some cleaning"

This reverts commit ef6074854cdede01b49121dfa626a61809b827d6.

babaq committed:
Revert "Revert "move stadardize ex into each prepare function, merge vlab getparam to base and some cleaning""

This reverts commit ab74a0edbb4f516c8a47e1ac5032c16be2642337.

babaq committed:
wrap up

babaq committed:
find and merge metadata changed to pass object reference instead of pass by value.  MetaTable cleanning, avoid of excessive copying.

Commit:
filename fix

babaq committed:
metatable syncrinize change to validate, dataexport do validate whenver export metadata

babaq committed:
Merge branch 'leo' of https://github.com/VLABSys/NeuroAnalysis into leo

Commit:
simplify filename

babaq committed:
vlab rough test pass

babaq committed:
Merge branch 'leo' of https://github.com/VLABSys/NeuroAnalysis into leo

babaq reviewed (commented):
@leoscholl  the code seems in a good shape. you can test with some visstim data, and then the metadata feature will be ready to merge into master.

babaq committed:
update about

Commit:
better factor parsing

Commit:
revert visstim name change

changing the filename for VisStim datasets will break the auto export function.. can't think of any workaround

babaq reviewed (dismissed):
ready to merge.

babaq committed:
Merge branch 'master' into leo
